sno/sno_context.py

```diff
@@ -0,0 +1,43 @@
+import click
+import pygit2
+import os
+
+
+class SnoContext(object):
+    def __init__(self):
+        self._repo_dir = None
+
+    @property
+    def repo_dir(self):
+        """The path of the repository. Defaults to the current directory."""
+        return self._repo_dir or os.curdir
+
+    @repo_dir.setter
+    def repo_dir(self, repo_dir):
+        self._repo_dir = repo_dir
+        if hasattr(self, "_repo"):
+            del self._repo
+
+    @property
+    def repo_dir_is_set(self):
+        """True if repo_dir has been explicitly set."""
+        return self._repo_dir is not None
+
+    @property
+    def repo(self):
```

Comment:
Can we add a `repo.delattr()` to allow us to deallocate the cached repo via `del ctx.obj.repo`? Windows is aggressive about file-locking, which potentially means we can't shell out to `git` with an open repo in all circumstances